### PR TITLE
⚡[FormState] remove needless renders

### LINKED ISSUE #183: [react-form-state] optimize rerenders
Right now the whole form rerenders whenever you change anything. We should be able to cache field objects intelligently such that this doesn't happen (assuming the fields are being rendered as pure components)

### PR DESCRIPTION
**Pull request by TheMallen**
closes #183 
Adds some memoization and other refactors to remove some needless re-renders, both on the field and whole render prop level.

### CODE REVIEW

File: packages/react-form-state/src/FormState.ts [hunk 1]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
-    return {
-      onChange: this.updateField.bind(this, path),
-      onBlur: this.blurField.bind(this, path),
-    };
+  @bind()
+  private fieldWithHandlers<Key extends keyof Fields>(
+    field: FieldStates<Fields>[Key],
+    fieldPath: Key,
+  ) {
+    return Object.assign({}, field, {
```
Comment (loic-d) [inline packages/react-form-state/src/FormState.ts]: Can you use the spread operator here?

File: packages/react-form-state/src/FormState.ts [hunk 2]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
+  }) {
+    const dirtyFieldsSet = new Set(dirtyFields);
+
+    if (dirty) {
+      dirtyFieldsSet.add(fieldPath);
+    } else {
+      dirtyFieldsSet.delete(fieldPath);
+    }
+
+    const newDirtyFields = Array.from(dirtyFieldsSet);
```
Comment (loic-d) [inline packages/react-form-state/src/FormState.ts]: Same here, can you use the spread operator `[...dirtyFieldSet]`?

File: packages/react-form-state/src/FormState.ts [hunk 3]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
+    dirty,
+    dirtyFields,
+  }: {
+    fieldPath: Key;
+    dirty: boolean;
+    dirtyFields: (keyof Fields)[];
+  }) {
+    const dirtyFieldsSet = new Set(dirtyFields);
+
+    if (dirty) {
```
Comment (loic-d) [inline packages/react-form-state/src/FormState.ts]: Nice to have this extracted into it's own function.

File: packages/react-form-state/src/FormState.ts [hunk 4]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
+    });
+
+    /*
+    We only want to update errors as the user types if they
+    already have an error.
+    https://polaris.shopify.com/patterns/error-messages#section-form-validation
+    */
+    const shouldUpdateError = field.error != null;
+    newField.error = shouldUpdateError
+      ? this.validateFieldValue(fieldPath, newField)
```
Comment (loic-d) [inline packages/react-form-state/src/FormState.ts]: Would it work with `newField.error = shouldUpdateError && this.validateFieldValue(fieldPath, newField)`?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+describe('<Nested />', () => {
+  beforeEach(() => {
+    faker.seed(ARBITRARY_SEED);
+  });
+
+  it('passes field state into child function', () => {
+    const renderPropSpy = jest.fn(() => null);
+
+    const product = {
+      title: faker.commerce.productName(),
```
Comment (loic-d) [inline packages/react-form-state/src/components/tests/Nested.tsx]: First time that I see `faker`, do we have access to it in `web`?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+              {renderPropSpy}
+            </FormState.Nested>
+          );
+        }}
+      </FormState>,
+    );
+
+    expect(renderPropSpy).toBeCalledWith({
+      title: {
+        name: 'nodejs.title',
```
Comment (loic-d) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Is this coming from `faker`?

File: packages/react-form-state/src/FormState.ts (hunk 1 above)
Comment (lemonmade) [reply packages/react-form-state/src/FormState.ts]: Agreed, even if you need to cast it is better to use the feature you want to use (also, apparently spread will be faster than object assign somehow /shrug)

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    mount(<FormState.Nested field={field}>{renderSpy}</FormState.Nested>);
+
+    const renderPropArgs = lastCallArgs(renderSpy);
+
+    expect(renderPropArgs.title.error).toBe(field.error.title);
+    expect(renderPropArgs.department.error).toBe(field.error.department);
+  });
+});
+
+function lastCallArgs(spy: jest.Mock) {
```
Comment (loic-d) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Should this function be extracted to a shared file since it is also used by `FormState`?

File: packages/react-form-state/src/FormState.ts [hunk 8]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
+    const dirtyFieldsSet = new Set(dirtyFields);
+
+    if (dirty) {
+      dirtyFieldsSet.add(fieldPath);
+    } else {
+      dirtyFieldsSet.delete(fieldPath);
+    }
+
+    const newDirtyFields = Array.from(dirtyFieldsSet);
+    return isEqual(dirtyFields, newDirtyFields) ? dirtyFields : newDirtyFields;
```
Comment (lemonmade) [inline packages/react-form-state/src/FormState.ts]: I think you can cheat and just use length here instead of a full comparison

File: packages/react-form-state/src/FormState.ts [hunk 9]
```diff
@@ -186,52 +180,103 @@ export default class FormState<
...
+    field: FieldStates<Fields>[Key];
+    value: Fields[Key];
+    dirty: boolean;
+  }) {
+    const newField = Object.assign({}, field, {
+      value,
+      dirty,
+    });
+
+    /*
```
Comment (lemonmade) [inline packages/react-form-state/src/FormState.ts]: This format of comments feels really odd — IMO line comments are better here

File: packages/react-form-state/src/FormState.ts (hunk 4 above)
Comment (lemonmade) [reply packages/react-form-state/src/FormState.ts]: Not really sure why you are mutating the object instead of making this a property this in the third argument to `newField`?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+import {mount} from 'enzyme';
+
+import {Input} from './components';
+import FormState from '../..';
+
+const ARBITRARY_SEED = 1337;
+
+describe('<Nested />', () => {
+  beforeEach(() => {
+    faker.seed(ARBITRARY_SEED);
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Why do you need to seed this?

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 6 above)
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Where does this come from?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    });
+  });
+
+  it("updates the top level FormState's array when an inner field is updated", () => {
+    const product = {title: faker.commerce.productName()};
+    const newTitle = faker.commerce.productName();
+
+    const renderPropSpy = jest.fn(({fields: {product}}: any) => {
+      return (
+        <>
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Don't need this?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+    const form = mount(
+      <FormState initialValues={{product}}>{renderPropSpy}</FormState>,
+    );
+
+    form
+      .find(Input)
+      .props()
+      .onChange(newTitle);
+    form.update();
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Why is this part needed?

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+    });
+
+    const form = mount(
+      <FormState initialValues={{product}}>{renderPropSpy}</FormState>,
+    );
+
+    form
+      .find(Input)
+      .props()
+      .onChange(newTitle);
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Bring in `trigger`?

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 5 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: yes!

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+    form.update();
+
+    const {fields} = lastCallArgs(renderPropSpy);
+
+    expect(fields.product.value.title).toBe(newTitle);
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: `expect(renderPropSpy).toHaveBeenLastCalledWith(expect.objectContaining(...))`

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        error: undefined,
+        initialValue: product.adjective,
+        value: product.adjective,
+        onChange: expect.any(Function),
+        onBlur: expect.any(Function),
+      },
+    });
+  });
+
+  it("updates the top level FormState's array when an inner field is updated", () => {
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Please use real apostrophes so you don't need to change the quote style

File: packages/react-form-state/src/components/tests/Nested.tsx [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    expect(renderPropArgs.department.error).toBe(field.error.department);
+  });
+});
+
+function lastCallArgs(spy: jest.Mock) {
+  const calls = spy.mock.calls;
+  return calls[calls.length - 1][0];
+}
+
+function TextField({onChange, ...props}: any) {
```
Comment (lemonmade) [inline packages/react-form-state/src/components/tests/Nested.tsx]: Need this and `Input`?

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 6 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: No, I see why you're confused though. This is coming from the keypath of the `title` field. `nodejs` seems to be an artifact of the test environment.

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 17]
```diff
@@ -124,6 +124,36 @@ describe('<FormState />', () => {
...
+      reset();
+      expect(fields).toMatchObject({
+        product: {
+          value: product,
+          initialValue: product,
+          dirty: false,
+        },
+      });
+    });
+  });
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: Are any of the above tests actually related to the changes you are making?

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 7 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: possibly

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 18]
```diff
@@ -652,8 +682,75 @@ describe('<FormState />', () => {
       expect(descriptionError.message).toEqual(submitErrors[1].message);
     });
   });
+
+  describe('performance', () => {
+    it('does not re-render form when state after onchange is identical', () => {
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: `onChange`

File: packages/react-form-state/src/FormState.ts (hunk 2 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: We can use spread here I think, though I'm not sure which reads nicer.

File: packages/react-form-state/src/FormState.ts (hunk 1 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: I avoided spread in cases where I would need to cast to satisfy typescript.

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 19]
```diff
@@ -652,8 +682,75 @@ describe('<FormState />', () => {
...
+      const renderPropSpy = jest.fn(() => null);
+
+      mount(
+        <FormState initialValues={{product, description}}>
+          {renderPropSpy}
+        </FormState>,
+      );
+
+      const formDetails = lastCallArgs(renderPropSpy);
+      formDetails.fields.product.onChange(product);
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: I think it would be nicer in the tests to have them simulate things a little better, so spread this field into an input and then trigger that input `onChange`

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 20]
```diff
@@ -652,8 +682,75 @@ describe('<FormState />', () => {
...
+
+  describe('performance', () => {
+    it('does not re-render form when state after onchange is identical', () => {
+      const product = faker.commerce.productName();
+      const description = faker.lorem.words();
+
+      const renderPropSpy = jest.fn(() => null);
+
+      mount(
+        <FormState initialValues={{product, description}}>
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: You don't need description, just distracts from the purpose of the test

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 21]
```diff
@@ -652,8 +682,75 @@ describe('<FormState />', () => {
...
+        </FormState>,
+      );
+
+      const formDetails = lastCallArgs(renderPropSpy);
+      formDetails.fields.product.onChange(product);
+
+      expect(renderPropSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('only renders renderProp once on mount', () => {
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: Not a sentence

File: packages/react-form-state/src/test/FormState.test.tsx [hunk 22]
```diff
@@ -652,8 +682,75 @@ describe('<FormState />', () => {
...
+      mount(
+        <FormState initialValues={{product, description}}>
+          {renderPropSpy}
+        </FormState>,
+      );
+
+      expect(renderPropSpy).toHaveBeenCalledTimes(1);
+    });
+
+    it('does not cause pureComponent re-renders when a different field is changed', () => {
```
Comment (lemonmade) [inline packages/react-form-state/src/test/FormState.test.tsx]: Your casing is all over the place :S also not a sentence

File: packages/react-form-state/src/FormState.ts (hunk 4 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: I don't think so because `shouldUpdateError` would be false not undefined.

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 5 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: It's pretty handy.

File: packages/react-form-state/src/FormState.ts (hunk 8 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: what if a new dirty field was added and one was removed?

File: packages/react-form-state/src/test/FormState.test.tsx (hunk 19 above)
Comment (TheMallen) [reply packages/react-form-state/src/test/FormState.test.tsx]: probably we actually want both cases

File: packages/react-form-state/src/FormState.ts (hunk 8 above)
Comment (lemonmade) [reply packages/react-form-state/src/FormState.ts]: Unless I am mistaken that is impossible in this case given that you only operate on a single keypath?

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 10 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: we can probably burn this

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 6 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: the name field is generated from the keypath, but for some reason in the test env it always includes this as the first segment

File: packages/react-form-state/src/test/FormState.test.tsx (hunk 17 above)
Comment (TheMallen) [reply packages/react-form-state/src/test/FormState.test.tsx]: they're just tests that were missing

File: packages/react-form-state/src/test/FormState.test.tsx (hunk 21 above)
Comment (TheMallen) [reply packages/react-form-state/src/test/FormState.test.tsx]: same

File: packages/react-form-state/src/FormState.ts (hunk 1 above)
Comment (loic-d) [reply packages/react-form-state/src/FormState.ts]: Yeah I had the same issues with the spread operator and casting

File: packages/react-form-state/src/FormState.ts (hunk 2 above)
Comment (loic-d) [reply packages/react-form-state/src/FormState.ts]: In Discounts we tried to use the `[...dirtyFieldSet]` syntax we we could, but it's a personal preference.

File: packages/react-form-state/src/FormState.ts (hunk 1 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: I guess we can just do it with spread, I just really don't like having casts all over the place as it feels messy and error prone.

File: packages/react-form-state/src/FormState.ts (hunk 8 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: 🤔 true

File: packages/react-form-state/src/FormState.ts (hunk 4 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: I actually don't remember why I mutated here, I'll take a look when I go over this this afternoon

Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: It was because of some awkwardness comparing the field object... Found a much nicer way to structure this though

File: packages/react-form-state/src/components/tests/Nested.tsx (hunk 14 above)
Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: I don't like that for single field comparisons when we already need `lastCallArgs` in this file anyway.

Comment (TheMallen) [reply packages/react-form-state/src/components/tests/Nested.tsx]: That said I should change some of the other places to do that.

File: packages/react-form-state/src/test/FormState.test.tsx (hunk 17 above)
Comment (TheMallen) [reply packages/react-form-state/src/test/FormState.test.tsx]: Actually I think they were partially some kind of rebase issue

File: packages/react-form-state/src/test/FormState.test.tsx (hunk 20 above)
Comment (TheMallen) [reply packages/react-form-state/src/test/FormState.test.tsx]: I do need it in this object since we are testing whether the description field rerenders. I can make this value just an empty string though.

File: packages/react-form-state/src/FormState.ts (hunk 2 above)
Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: I'm leaving this for now, seems like there's a missing transpilation issue or something, and I kind of like how this reads anyway.

File: packages/react-form-state/src/tests/FormState.test.tsx [hunk 23]
```diff
@@ -105,15 +101,37 @@ describe('<FormState />', () => {
...
       formDetails.fields.product.onChange(faker.commerce.productName());
-      form.update();
 
       const {reset} = lastCallArgs(renderPropSpy);
       reset();
 
-      form.update();
-
       const {fields} = lastCallArgs(renderPropSpy);
+      expect(fields).toMatchObject({
```
Comment (TheMallen) [inline packages/react-form-state/src/tests/FormState.test.tsx]: I left a lot of these using `lastCallArgs` because we only actually care about a few of the fields, and doing  `expect.objectContaining` with `toHaveLastBeenCalledWith` adds too much noise and stuff we don't care about IMO. 

I did use that strategy in places we do actually care about every field though.

File: packages/react-form-state/src/tests/FormState.test.tsx [hunk 24]
```diff
@@ -156,17 +174,14 @@ describe('<FormState />', () => {
       const formDetails = lastCallArgs(renderPropSpy);
       formDetails.fields.product.onChange(faker.commerce.productName());
       formDetails.fields.color.onChange(faker.commerce.color());
-      form.update();
```
Comment (TheMallen) [inline packages/react-form-state/src/tests/FormState.test.tsx]: found and replaced all of these because they weren't needed

File: packages/react-form-state/src/FormState.ts [hunk 25]
```diff
@@ -186,52 +180,106 @@ export default class FormState<
...
-  private getHandlersForField(path: keyof FieldStates<Fields>) {
+  @bind()
+  private fieldWithHandlers<Key extends keyof Fields>(
+    field: FieldStates<Fields>[Key],
+    fieldPath: Key,
+  ) {
     return {
-      onChange: this.updateField.bind(this, path),
-      onBlur: this.blurField.bind(this, path),
+      ...(field as FieldState<Fields[Key]>),
```
Comment (tzvipm) [inline packages/react-form-state/src/FormState.ts]: do we not need the `name: path` piece any more?

Comment (TheMallen) [reply packages/react-form-state/src/FormState.ts]: Added back
